[18.0][MIG] sale_procurement_group_by_line: Migration to 18.0.

MeritxellAForgeFlow opened the pull request:
Migration to 18.0

@ForgeFlow

yvaucher committed:
Add module sale_procurement_group_by_line

This module extract logic to create multiple procurement group
for a single sale order for grouped sale order lines.

yvaucher committed:
Add OCA module icon

yvaucher committed:
Add bug tracker link on README.rst

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

Commit:
Recreate pickings doesn't show View Delivery Orders button

* Fix #264
* Fix : Can't create delivery from shipping exception

pedrobaeza committed:
[MIG] Rename manifest files

serpentcs-dev1 committed:
9.0 sale_procurement_group_by_line

LoisRForgeFlow committed:
sale_delivery_block (#434)

* [ADD] [9.0] sale_delivery_block

* [IMP] Add data with an example Delivery Block Reason.

* [FIX]
 * Able to edit the Delivery Block Reason in state 'sent'
 * README

* [FIX] make sale_delivery_block compatible with sale_procurement_group_by_line

* [FIX] travis and data nouptade.

* [ADD] delivery block tests.

* [IMP] Add the 'Default Delivery Block Reason' in partners.

* [IMP] track visibility of delivery block reason

* [9.0][IMP] sale_delivery_block:
  * default_delivery_block is now a comercial field.
  * When duplicating a sale order the delivery block is recomputed.

* [9.0][UPD] sale_delivery_block_proc_group_by_line: Update README.

* [FIX] api.model used with api.constrains

MiquelRForgeFlow committed:
[MIG] sale_procurement_group_by_line: Migration to 10.0

AaronHForgeFlow committed:
[FIX]get picking on sale_procurement_group_by_line

AaronHForgeFlow committed:
[FIX]adapt test to match

murtuzasaleh committed:
[11.0] [MIG]sale_procurement_group_by_line: Migration to 11.0.

MiquelRForgeFlow committed:
[IMP] Restore super in _action_launch_procurement_rule

marylla committed:
Translated using Weblate (German)

Currently translated at 66.7% (2 of 3 strings)

Translation: sale-workflow-11.0/sale-workflow-11.0-sale_procurement_group_by_line
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-11-0/sale-workflow-11-0-sale_procurement_group_by_line/de/

MiquelRForgeFlow committed:
[MIG] sale_procurement_group_by_line: Migration to 12.0

JoanSForgeFlow committed:
[IMP] sale_procurement_group_by_line: black, isort

JoanSForgeFlow committed:
[MIG] sale_procurement_group_by_line: Migration to 13.0

MiquelRForgeFlow committed:
*: E741 ambiguous variable name 'l'

danimv5 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (3 of 3 strings)

Translation: sale-workflow-13.0/sale-workflow-13.0-sale_procurement_group_by_line
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-13-0/sale-workflow-13-0-sale_procurement_group_by_line/es/

LoisRForgeFlow committed:
[13.0][FIX] sale_procurement_group_by_line: fill previous_product_uom_qty.

This avoids duplicated procurements in some cases, basically for
any operation that has a specific `<sale.order.line>._get_qty_procurement`
relying on `previous_product_uom_qty`. E.g: manufacturing kits,
dropshipping...

koenloodts committed:
[14.0-mig-sale] [IMP] sale_procurement_group_by_line: black, isort, prettier

koenloodts committed:
[14.0-mig-sale] [MIG] sale_procurement_group_by_line: Migration to 14.0

JordiMForgeFlow committed:
[13.0][FIX]sale_procurement_group_by_line: fix fill previous_product_uom_qty

Prevents having duplicated procurements when having several order lines involving any operation that has a specific `<sale.order.line>._get_qty_procurement`
relying on `previous_product_uom_qty` (Manufacturing kits, dropshipping..)

JordiMForgeFlow committed:
[13.0][FIX]sale_procurement_group_by_line: use SO company as the procurement company

JasminSForgeFlow committed:
[IMP] sale_procurement_group_by_line: black, isort, prettier

JasminSForgeFlow committed:
[MIG] sale_procurement_group_by_line: Migration to 15.0

Noel000 committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (6 of 6 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_procurement_group_by_line
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_procurement_group_by_line/ca/

JasminSForgeFlow committed:
[IMP] sale_procurement_group_by_line: Update Devlopment Status

JordiMForgeFlow committed:
[FIX] sale_procurement_group_by_line: fix fill previous_product_uom_qty

Before the fix, when new quantities are added into an already confirmed
line, we are writing in the previous_product_uom_qty only the new qty that
has been processed, not the total amount that has actually been
processed for the line.

Due to this, when we call super() on _action_launch_stock_rule, the new
quantities are procured again, as we have not really updated the amount
in previous_product_uom_qty including what we have already procured
before calling super().

AaronHForgeFlow committed:
[FIX] sale_procurement_group_by_line: group negative moves when reducing qty in a sales order

MiquelRForgeFlow committed:
[FIX] sale_procurement_group_by_line: adapt to v15

MiquelRForgeFlow committed:
[IMP] sale_procurement_group_by_line: add test to avoid regression

Commit:
[MIG] sale_procurement_group_by_line: Migration to 16.0

Commit:
[UPD] Update sale_procurement_group_by_line.pot

mymage committed:
Translated using Weblate (Italian)

Currently translated at 33.3% (1 of 3 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_procurement_group_by_line
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_procurement_group_by_line/it/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (3 of 3 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_procurement_group_by_line
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_procurement_group_by_line/es/

dbruehlmeier committed:
Translated using Weblate (German)

Currently translated at 100.0% (3 of 3 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_procurement_group_by_line
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_procurement_group_by_line/de/

celm1990 committed:
[IMP] sale_procurement_group_by_line: pre-commit auto fixes

celm1990 committed:
[MIG] sale_procurement_group_by_line: Migration to 17.0

Commit:
[UPD] Update sale_procurement_group_by_line.pot

OCA-git-bot committed:
[BOT] post-merge updates

MeritxellAForgeFlow committed:
[MIG] sale_procurement_group_by_line: Migration to 18.0

JordiMForgeFlow reviewed (commented):
Could you include the module name in the MIG commit like: [MIG] sale_procurement_group_by_line: Migration to V18?

Code review LGTM :)

RicardCForgeFlow reviewed (approved):
Code review :+1:

MiquelRForgeFlow commented on sale_procurement_group_by_line/model/sale.py:
```diff
@@ -0,0 +1,119 @@
+            previous_product_uom_qty = {}
+        for line in self:
+            line = line.with_company(line.company_id)
+            if line.state != "sale" or line.product_id.type not in ("consu", "product"):
```
```suggestion
            if line.state != "sale" or line.order_id.locked or line.product_id.type != 'consu':
```

AaronHForgeFlow reviewed (approved):
Code review LGTM

MiquelRForgeFlow commented on sale_procurement_group_by_line/model/sale.py:
```diff
@@ -0,0 +1,119 @@
+
+            line_uom = line.product_uom
+            quant_uom = line.product_id.uom_id
+            product_qty, procurement_uom = line_uom._adjust_uom_quantities(
+                product_qty, quant_uom
+            )
+
+            procurements.append(
+                self.env["procurement.group"].Procurement(
+                    line.product_id,
+                    product_qty,
+                    procurement_uom,
+                    line.order_id.partner_shipping_id.property_stock_customer,
+                    line.display_name,
+                    line.order_id.name,
+                    line.order_id.company_id,
+                    values,
+                )
+            )
```
```suggestion
            origin = f'{line.order_id.name} - {line.order_id.client_order_ref}' if line.order_id.client_order_ref else line.order_id.name
            product_qty, procurement_uom = line_uom._adjust_uom_quantities(
                product_qty, quant_uom
            )
            procurements += line._create_procurements(product_qty, procurement_uom, origin, values)
```

MiquelRForgeFlow commented on sale_procurement_group_by_line/tests/test_sale_procurement_group_by_line.py:
```diff
@@ -0,0 +1,144 @@
+        self.sale.action_confirm()
+        self.assertEqual(self.sale.state, "sale")
+        self.assertEqual(len(self.line1.move_ids), 1)
+        self.assertEqual(self.line1.move_ids.name, self.line1.display_name)
+        self.assertEqual(len(self.line2.move_ids), 1)
+        self.assertEqual(self.line2.move_ids.name, self.line2.display_name)
```
```suggestion
        self.assertEqual(self.line1.move_ids.name, self.line1.product_id.display_name)
        self.assertEqual(len(self.line2.move_ids), 1)
        self.assertEqual(self.line2.move_ids.name, self.line2.product_id.display_name)
```
or something like this.

It seems it's due to https://github.com/odoo/odoo/commit/4cc2b3bcca613dfb37bb77e5fef6de6e044e3399

MiquelRForgeFlow committed:
[FIX] sale_procurement_group_by_line: update to odoo

Adapt to:

- https://github.com/odoo/odoo/commit/4cc2b3bcca613dfb37bb77e5fef6de6e044e3399.
- https://github.com/odoo/odoo/commit/17bece3e797913bcba8dd7e07fc8541c0a45e3f7
- https://github.com/odoo/odoo/commit/60f08ffbccca07adba6a5449d22496b14c321901

MiquelRForgeFlow commented:
Added v17 changes in a separated commit (a forward port of https://github.com/OCA/sale-workflow/pull/3428).

RicardCForgeFlow commented:
Hi @rousseldenis, could you check this PR? Thank you very much!  :smile:

rousseldenis commented:
/ocabot migration sale_procurement_group_by_line

rousseldenis commented on sale_procurement_group_by_line/model/sale.py:
```diff
@@ -43,7 +43,7 @@ def _action_launch_stock_rule(self, previous_product_uom_qty=False):
                 line.state != "sale"
                 or line.order_id.locked
-                or line.product_id.type not in ("consu", "product")
+                or line.product_id.type != "consu"
```
I'm wondering if we should test `is_storable` field

RicardCForgeFlow replied:
Hi @rousseldenis, I see Odoo does this

[sale_stock/models/sale_order_line.py#L357](https://github.com/odoo/odoo/blob/be0d490f6f541a965ce8ca6d5a395f56fc6ad8cf/addons/sale_stock/models/sale_order_line.py#L357)

Although is_storable=False implies no inventory tracking, consumable products (type: consu, is_storable: False) may still require provisioning rules.

Do you agree?

rvalyi reviewed (approved):
code review LGTM

rvalyi commented:
/ocabot merge nobump
